File: problems/medium/uniq_path.rb
# https://leetcode.com/problems/unique-paths/
# @param {Integer} m
# @param {Integer} n
# @return {Integer}
def unique_paths(m, n)
  return 1 if (m == 0 || n == 0)
  i=0
  # define the state
  # f is the possible uniq paths that arrive at this point. 
  f = []
  while (i<m)
    f.push({})
    i += 1
  end
  # initalize
  f[0][0] = 1
  i = 0
  while(i<m)
    f[i][0] = 1
    i += 1
  end
  i = 0
  while(i<n)
    f[0][i] = 1
    i += 1
  end
  # function
  i = 1
  while(i<m)
    j = 1
    while(j<n)
      f[i][j] = f[i-1][j] + f[i][j-1]
      j += 1
    end
    i += 1
  end
  # result
  f[m-1][n-1]
end

File: type/dp/uniq_path.go
/*
 * @Author: hongwei.sun
 * @Date: 2021-01-22 18:45:51
 * @LastEditors: your name
 * @LastEditTime: 2024-03-12 23:31:42
 * @Description: file content
 */
// https://leetcode.com/problems/unique-paths/
/*
A robot is located at the top-left corner of a m x n grid
(marked 'Start' in the diagram below).

The robot can only move either down or right at any point in time.
The robot is trying to reach the bottom-right corner of the grid
(marked 'Finish' in the diagram below).

How many possible unique paths are there?


Above is a 7 x 3 grid. How many possible unique paths are there?

Note: m and n will be at most 100.

Example 1:

Input: m = 3, n = 2
Output: 3
Explanation:
From the top-left corner, there are a total of 3 ways to reach
the bottom-right corner:
1. Right -> Right -> Down
2. Right -> Down -> Right
3. Down -> Right -> Right
Example 2:

Input: m = 7, n = 3
Output: 28
*/
// use DP
func uniquePaths(m int, n int) int {
	// f[i][j] is the uniq path amount when it arrives i and j from [0][0]
	f := make([][]int, m)
	for i := 0; i < m; i++ {
		f[i] = make([]int, n)
	}
	for i := 0; i < m; i++ {
		f[i][0] = 1
	}
	for i := 0; i < n; i++ {
		f[0][i] = 1
	}
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			// when it is going to arrive [i][j], two possible position where it is now:
			// [i-1][j] or [i][j-1]
			f[i][j] = f[i-1][j] + f[i][j-1]
		}
	}
	return f[m-1][n-1]
}
